sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/EstimationUtils.scala

```diff
@@ -114,4 +115,183 @@ object EstimationUtils {
+            // --------+------------------+------------+----------------+------->
+            val leftRatio = (right.hi - left.lo) / (left.hi - left.lo)
+            val rightRatio = (right.hi - left.lo) / (right.hi - right.lo)
+            if (leftRatio == 0) {
+              // The overlapped range has only one value.
+              OverlappedRange(
+                lo = right.hi,
+                hi = right.hi,
+                leftNdv = 1,
+                rightNdv = 1,
+                leftNumRows = leftHeight / left.ndv,
+                rightNumRows = rightHeight / right.ndv
+              )
+            } else {
+              OverlappedRange(
+                lo = left.lo,
+                hi = right.hi,
+                leftNdv = left.ndv * leftRatio,
+                rightNdv = right.ndv * rightRatio,
+                leftNumRows = leftHeight * leftRatio,
+                rightNumRows = rightHeight * rightRatio
+              )
+            }
+          } else if (right.lo >= left.lo && right.hi <= left.hi) {
+            // Case5: the left bin contains the right bin
+            //      left.lo            right.lo     right.hi         left.hi
+            // --------+------------------+------------+----------------+------->
+            val leftRatio = (right.hi - right.lo) / (left.hi - left.lo)
+            OverlappedRange(
+              lo = right.lo,
+              hi = right.hi,
+              leftNdv = left.ndv * leftRatio,
+              rightNdv = right.ndv,
+              leftNumRows = leftHeight * leftRatio,
+              rightNumRows = rightHeight
+            )
+          } else {
+            assert(right.lo <= left.lo && right.hi >= left.hi)
+            // Case6: the right bin contains the left bin
+            //      right.lo           left.lo      left.hi          right.hi
+            // --------+------------------+------------+----------------+------->
+            val rightRatio = (left.hi - left.lo) / (right.hi - right.lo)
+            OverlappedRange(
+              lo = left.lo,
+              hi = left.hi,
+              leftNdv = left.ndv,
+              rightNdv = right.ndv * rightRatio,
+              leftNumRows = leftHeight,
+              rightNumRows = rightHeight * rightRatio
+            )
+          }
+          overlappedRanges += range
+        }
+      }
+    }
+    overlappedRanges
+  }
+
+  /**
+   * Given an original bin and a value range [min, max], returns the trimmed bin and its number of
+   * rows.
+   */
+  def trimBin(bin: HistogramBin, height: Double, min: Double, max: Double)
+    : (HistogramBin, Double) = {
+    val (lo, hi) = if (bin.lo <= min && bin.hi >= max) {
+      //       bin.lo              min          max          bin.hi
+      // --------+------------------+------------+-------------+------->
+      (min, max)
+    } else if (bin.lo <= min && bin.hi >= min) {
+      //       bin.lo              min        bin.hi
```

Comment:
what if the `max` is after the `bin.hi`?

Reply:
in this case, `max` is after the `bin.hi`, so the trimmed part is `(min, bin.hi)`. I'll update the figure to indicate that.